keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+  but supporting sharing between pods would need a completely new
+  API similar to the one in this KEP.
+- Using a resource that is expensive to initialize multiple times
+  in different pods. This is not possible at the moment.
+- Custom parameters that describe resource requirements and initialization.
+  Parameters are not limited to a single, linear quantity that can be counted.
+  With the current Pod API, annotations have to be used to capture such
+  parameters and then hacks are needed to access them from a CSI driver or
+  device plugin.
+
+Support for new hardware will be provided by hardware vendor add-ons. It will
+not be necessary anymore to modify Kubernetes itself.
+
+This KEP does not replace other means of requesting traditional resources
+(RAM/CPU, volumes, extended resources). The scheduler will serve as coordinator
+between the add-ons which own resources (CSI driver, resource driver) and the
+resources owned and assigned by the scheduler (RAM/CPU, extended resources).
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Originally, Kubernetes and its scheduler only tracked CPU and RAM as
+resources for containers. Later, support for storage and discrete,
+countable per-node extended resources was added. The kubelet device plugin
+interface then made such local resources available to containers. But
+for many newer devices, this approach and the Kubernetes API for
+requesting these custom resources is too limited. This KEP addresses
+limitations of the current approach for the following use cases:
+
+- *Device initialization*: When starting a workload that uses
+  an accelerator like an FPGA, I’d like to have the accelerator
+  reconfigured or reprogrammed for the workload before the workload
+  itself starts. For security reasons, workloads should not be able to
+  reconfigure devices directly.
+
+  *Limitation*: Currently, it’s impossible to specify the desired
+  device properties that are required for reconfiguring devices.
+  For the FPGA example, a file containing the desired configuration
+  of the FPGA has to be referenced.
+
+- *Device cleanup*: When my workload is finished, I would like to have
+  a mechanism for cleanup of the device, that will ensure that device
+  does not contain traces/parameters/data from previous workloads and
+  appropriate power state/shutdown. For example, an FPGA might have
+  to be reset because its configuration for the workload was
+  confidential.
+
+  *Limitation*: Post-stop actions are not supported.
+
+- *Partial allocation*: When deploying a container I’d like to be able
```

Comment:
I think this was meant to refer to virtual devices provided by the same hardware.

@klueska: NVIDIA cards support this, right? Can you perhaps suggest better wording and a specific example?

Reply:
I think there are three things relevant to NVIDIA cards here.

1. Dynamic provisioning of MIG devices based on user-provided workload constraints (i.e. hard partitioning of memory and compute resources in fixed size chunks). Right now, MIG devices have to be pre-configured and countable so that the scheduler can treat them as allocatable resources. With this new proposal, MIG devices could be created on the fly to service containers with specific GPU resource requirments.

2. Dynamic provisioning of MPS clients based on user-provided workload constraints (i.e. hard partitioning of memory in arbitrarily-sized chunks, but limited partitioning of compute resources). Right now we don't have a good way of supporting the MPS model in Kubernetes and have avoided officially supporting it for years. This new proposal will give us a path towards doing so now.

3. Providing shared (mostly unconstrained) access to a full GPU from multiple, cooperative containers and letting CUDA time-slice them under the hood. We support this today in a somewhat awkward fashion by creating a set of "replicas" for a single GPU and advertising them as separate devices to the scheduler. This acts as a proxy for providing shared access to the same GPU, but the interface is awkward and somewhat unintuitive (i.e. asking for 2 replicas doesn't give you proportionally more time on the GPU, you have just requested shared access to the GPU twice). The new prosal will allow us to support this use case in a more intuitive and well-defined way.

Reply:
ISTR that MIG re-config was heavyweight and required running workloads to be stopped - am I misremembering?

And end-to-end example showing a hypothetical MIG use-case would be great (since I have some context on it, but not a ton).

Reply:
> ISTR that MIG re-config was heavyweight and required running workloads to be stopped - am I misremembering?

Not quite. Getting a GPU into MIG mode is heavy-weight, but then reconfiguring it once the GPU is in MIG mode is very quick. With this proposal you could imagine a plugin advertising the possible set of MIG devices that it is capable of configuring and then creating them on the fly as they are requested.

> And end-to-end example showing a hypothetical MIG use-case would be great (since I have some context on it, but not a ton).

I will put together some text for an example later today or in the morning.

Reply:
Please see https://github.com/kubernetes/enhancements/pull/3064/files#r904855097 for some example text.

Reply:
I have added that as a separate commit to a pending update for this KEP, see https://github.com/pohly/enhancements/pull/16.

@thockin is that sufficient to address this comment (once moved, of course)?